src/hotspot/share/jvmci/jvmciCompilerToVM.cpp

```diff
@@ -620,8 +621,46 @@ C2V_VMENTRY_NULL(jobject, lookupClass, (JNIEnv* env, jobject, jclass mirror))
 
 C2V_VMENTRY_NULL(jobject, resolvePossiblyCachedConstantInPool, (JNIEnv* env, jobject, jobject jvmci_constant_pool, jint index))
   constantPoolHandle cp(THREAD, JVMCIENV->asConstantPool(jvmci_constant_pool));
-  oop result = cp->resolve_possibly_cached_constant_at(index, CHECK_NULL);
-  return JVMCIENV->get_jobject(JVMCIENV->get_object_constant(result));
+  oop obj = cp->resolve_possibly_cached_constant_at(index, CHECK_NULL);
+  constantTag tag = cp->tag_at(index);
+  if (tag.is_dynamic_constant() || tag.is_dynamic_constant_in_error()) {
+    if (obj == Universe::the_null_sentinel()) {
+      return JVMCIENV->get_jobject(JVMCIENV->get_JavaConstant_NULL_POINTER());
+    }
+    BasicType bt = Signature::basic_type(cp->uncached_signature_ref_at(index));
+    if (!is_reference_type(bt)) {
+      if (!is_java_primitive(bt)) {
+        return JVMCIENV->get_jobject(JVMCIENV->get_JavaConstant_ILLEGAL());
+      }
+      jvalue value;
+      jlong raw_value;
+      BasicType bt2 = java_lang_boxing_object::get_value(obj, &value);
+      assert(bt2 == bt, "");
+      switch (bt2) {
+        case T_BOOLEAN: raw_value = value.z; break;
+        case T_BYTE:    raw_value = value.b; break;
+        case T_SHORT:   raw_value = value.s; break;
+        case T_CHAR:    raw_value = value.c; break;
+        case T_INT:     raw_value = value.i; break;
+        case T_LONG:    raw_value = value.j; break;
+        case T_FLOAT: {
+          JVMCIObject result = JVMCIENV->call_JavaConstant_forFloat(value.f, JVMCI_CHECK_NULL);
+          return JVMCIENV->get_jobject(result);
+        }
+        case T_DOUBLE: {
+          JVMCIObject result = JVMCIENV->call_JavaConstant_forDouble(value.d, JVMCI_CHECK_NULL);
+          return JVMCIENV->get_jobject(result);
+        }
+        default: {
+          return JVMCIENV->get_jobject(JVMCIENV->get_JavaConstant_ILLEGAL());
+        }
+      }
+
+      JVMCIObject result = JVMCIENV->call_PrimitiveConstant_forTypeChar(type2char(bt2), raw_value, JVMCI_CHECK_NULL);
```

Comment:
Given the above switch is total this part is unreachable? Did you mean for it to occur after the `if (!is_reference_type(bt)) ` block?

Reply:
No - it should not be there at all: https://github.com/openjdk/jdk/pull/3578/commits/b70c7b1a64d51fce23edde860f04749e8e49d34f
Thanks for spotting.

Reply:
Oops, i think i need glasses, the switch has break statements that results in processing of the `raw_value`:
```c++
      switch (bt2) {
        case T_BOOLEAN: raw_value = value.z; break;
...
```

Reply:
I need more than glasses since I wrote this code! Thankfully the test I added also showed the problem with removing this "unreachable" code.